pkg/kubelet/cpumanager/cpu_manager.go

```diff
@@ -0,0 +1,363 @@
+
+	IsUnderCPUPressure() bool
+
+	lifecycle.PodAdmitHandler
+}
+
+type manager struct {
+	sync.Mutex
+	policy Policy
+	state  state.State
+
+	// containerRuntime is the container runtime service interface needed
+	// to make UpdateContainerResources() calls against the containers.
+	containerRuntime runtimeService
+
+	// kletGetter provides a method for listing all the pods on the node
+	// so all the containers can be updated in the reconciliation loop.
+	kletGetter kletGetter
+
+	// podStatusProvider provides a method for obtaining pod statuses
+	// and the containerID of their containers
+	podStatusProvider status.PodStatusProvider
+
+	// getPodsFunc gets list of active pods to consider for eviction when
+	// under pressure.
+	getPodsFunc eviction.ActivePodsFunc
+
+	// killPodFunc kills a pod selected for eviction when under pressure.
+	killPodFunc eviction.KillPodFunc
+
+	// recorder records eviction events.
+	recorder record.EventRecorder
+}
+
+var _ Manager = &manager{}
+
+// NewManager creates new cpu manager based on provided policy
+func NewManager(
+	cpuPolicyName string,
+	cr runtimeService,
+	kletGetter kletGetter,
+	statusProvider status.PodStatusProvider,
+	capacityProvider eviction.CapacityProvider,
+	getPodsFunc eviction.ActivePodsFunc,
+	killPodFunc eviction.KillPodFunc,
+	recorder record.EventRecorder,
+) (Manager, lifecycle.PodAdmitHandler, error) {
+	var policy Policy
+
+	switch policyName(cpuPolicyName) {
+	case PolicyNone:
+		policy = NewNonePolicy()
+	case PolicyStatic:
+		machineInfo, err := kletGetter.GetCachedMachineInfo()
+		if err != nil {
+			return nil, nil, err
+		}
+		topo, err := topology.Discover(machineInfo)
+		if err != nil {
+			return nil, nil, err
+		}
+		glog.Infof("[cpumanager] detected CPU topology: %v", topo)
+
+		resources := capacityProvider.GetNodeAllocatableReservation()
+		cpuResource, ok := resources[v1.ResourceCPU]
+		if ok {
+			reservedCores := int(cpuResource.Value())
+			glog.Infof("[cpumanager] reserving %v cores due to kube/system reserved", reservedCores)
+			topo.NumReservedCores = reservedCores
+		}
+		policy = NewStaticPolicy(topo)
+	default:
+		glog.Warningf("[cpumanager] Unknown policy (\"%s\"), falling back to \"%s\" policy (\"%s\")", cpuPolicyName, PolicyNone)
+		policy = NewNonePolicy()
+	}
+
+	manager := &manager{
+		policy:            policy,
+		state:             state.NewMemoryState(),
+		containerRuntime:  cr,
+		kletGetter:        kletGetter,
+		podStatusProvider: statusProvider,
+		getPodsFunc:       getPodsFunc,
+		killPodFunc:       killPodFunc,
+		recorder:          recorder,
+	}
+	return manager, manager, nil
+}
+
+func (m *manager) Start() {
+	glog.Infof("[cpumanger] starting with %s policy", m.policy.Name())
+	m.policy.Start(m.state)
+	if m.policy.Name() == string(PolicyNone) {
+		return
+	}
+	go wait.Until(func() { m.reconcileState() }, time.Second, wait.NeverStop)
+}
+
+func (m *manager) RegisterContainer(p *v1.Pod, c *v1.Container, containerID string) error {
```

Comment:
This workflow isn't obvious to me. To help me catch up can you describe (preferably as comments in the PR), when this method is expected to be invoked in the pod lifecycle and why this method has to invoke the UpdateResources method on the CRI, instead of say for example the pod lifecycle handler in kubelet invoking the CRI interfaces?

Reply:
Having the overall kubelet interactions documented will be very helpful. TBH, I find it hard to page in interactions between different modules for this feature.